mantidproject/mantid
# Muon - Create Fitting Contexts

**Description of work.**
This PR does some refactoring to the muon fitting tab. Any data that will potentially need saving in the future for project save is now stored in a context.

These contexts are lightweight and are only used to store the important data, while the model is still used for the heavy processing.

**To test:**
Test Single Fitting, Simultaneous fitting and TF Asymmetry fitting

Single Fitting:

1. Open Muon Analysis interface
2. Load MUSR `62260-2`
3. Go to Fitting tab
4. Add a `GausOsc` function
5. Click `Fit` and see that the Fit Status and Chi Squared changes.
6. Go to the next dataset using `>>`.
7. See the Fit Status and Chi Squared are empty. Click `Fit`. The Fit Status and Chi Squared should change.
8. Switch to previous fit using `<<` and see that the Fit Status and chi squared are different for each fit.
9. Click `Undo Fit`. This will undo the previous fit, so go back to the latest fit using `>>` and the Fit Status and Chi Squared should be empty.

Simultaneous Fitting:

1. Open Muon Analysis interface
2. Load MUSR `62260-2`
3. Go to Fitting tab
4. Add a `GausOsc` function
5. Tick Simultaneous over
6. Change Frequency parameter to 1.3 and tick global.
7.  Click `Fit`. Switch between all the datasets using `>>` 
8. Check that Chi Squared is the same for all datasets
9. Check that the Frequency parameter is the same for all datasets.

TF Asymmetry fitting:

1. Open Muon Analysis interface
2. Load MUSR `62260-2`
3. Go to Fitting tab
4. Switch Fitting type to `TF Asymmetry`
4. Add a `GausOsc` function
5. Tick Simultaneous over
6. Change Frequency parameter to 1.3 and tick global.
7.  Click `Fit`. Switch between all the datasets using `>>` 
8. Check that Chi Squared is the same for all datasets
9. Check that the Frequency parameter is the same for all datasets.

Fixes #31445 

*This does not require release notes* because **this is refactoring**

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
